File: src/main/java/com/weeth/domain/attendance/application/mapper/AttendanceMapper.java
package com.weeth.domain.attendance.application.mapper;

import com.weeth.domain.attendance.application.dto.AttendanceDTO;
import com.weeth.domain.attendance.domain.entity.Attendance;
import com.weeth.domain.user.domain.entity.User;
import org.mapstruct.*;

import java.util.List;

@Mapper(componentModel = MappingConstants.ComponentModel.SPRING, unmappedTargetPolicy = ReportingPolicy.IGNORE)
public interface AttendanceMapper {

    @Mappings({
            @Mapping(target = "attendanceRate", source = "user.attendanceRate"),
            @Mapping(target = "title", source = "attendance.meeting.title"),
            @Mapping(target = "status", source = "attendance.status"),
            @Mapping(target = "code", ignore = true),
            @Mapping(target = "start", source = "attendance.meeting.start"),
            @Mapping(target = "end", source = "attendance.meeting.end"),
            @Mapping(target = "location", source = "attendance.meeting.location"),
    })
    AttendanceDTO.Main toMainDto(User user, Attendance attendance);

    @Mappings({
            @Mapping(target = "attendanceRate", source = "user.attendanceRate"),
            @Mapping(target = "title", source = "attendance.meeting.title"),
            @Mapping(target = "status", source = "attendance.status"),
            @Mapping(target = "code", source = "attendance.meeting.code"),
            @Mapping(target = "start", source = "attendance.meeting.start"),
            @Mapping(target = "end", source = "attendance.meeting.end"),
            @Mapping(target = "location", source = "attendance.meeting.location"),
    })
    AttendanceDTO.Main toAdminResponse(User user, Attendance attendance);

    @Mappings({
            @Mapping(target = "attendances", source = "attendances"),
            @Mapping(target = "total", expression = "java( user.getAttendanceCount() + user.getAbsenceCount() )")
    })
    AttendanceDTO.Detail toDetailDto(User user, List<AttendanceDTO.Response> attendances);

    @Mappings({
            @Mapping(target = "title", source = "attendance.meeting.title"),
            @Mapping(target = "start", source = "attendance.meeting.start"),
            @Mapping(target = "end", source = "attendance.meeting.end"),
            @Mapping(target = "location", source = "attendance.meeting.location"),
    })    AttendanceDTO.Response toResponseDto(Attendance attendance);

    @Mappings({
            @Mapping(target = "id", source = "attendance.id"),
            @Mapping(target = "status", source = "attendance.status"),
            @Mapping(target = "name", source = "attendance.user.name"),
            @Mapping(target = "position", source = "attendance.user.position"),
            @Mapping(target = "department", source = "attendance.user.department"),
            @Mapping(target = "studentId", source = "attendance.user.studentId")
    })
    AttendanceDTO.AttendanceInfo toAttendanceInfoDto(Attendance attendance);

}

File: src/main/kotlin/com/weeth/domain/attendance/application/mapper/AttendanceMapper.kt
package com.weeth.domain.attendance.application.mapper

import com.weeth.domain.attendance.application.dto.response.AttendanceDetailResponse
import com.weeth.domain.attendance.application.dto.response.AttendanceInfoResponse
import com.weeth.domain.attendance.application.dto.response.AttendanceResponse
import com.weeth.domain.attendance.application.dto.response.AttendanceSummaryResponse
import com.weeth.domain.attendance.application.dto.response.QrTokenResponse
import com.weeth.domain.attendance.domain.entity.Attendance
import com.weeth.domain.attendance.domain.enums.AttendanceStatus
import com.weeth.domain.club.domain.entity.ClubMember
import com.weeth.domain.club.domain.vo.ClubAttendanceStats
import com.weeth.domain.session.domain.entity.Session
import org.springframework.stereotype.Component
import java.time.LocalDateTime

@Component
class AttendanceMapper {
    fun toSummaryResponse(
        clubMember: ClubMember,
        attendance: Attendance?,
    ): AttendanceSummaryResponse =
        AttendanceSummaryResponse(
            attendanceRate = clubMember.attendanceStats.attendanceRate,
            title = attendance?.session?.title,
            status = attendance?.status,
            sessionId = attendance?.session?.id,
            start = attendance?.session?.start,
            end = attendance?.session?.end,
            location = attendance?.session?.location,
        )

    fun toDetailResponse(
        cardinalNumber: Int,
        attendances: List<Attendance>,
    ): AttendanceDetailResponse {
        val stats = toStats(attendances)
        return AttendanceDetailResponse(
            attendanceCount = stats.attendanceCount,
            total = stats.attendanceCount + stats.absenceCount,
            absenceCount = stats.absenceCount,
            attendances = attendances.map(::toResponse),
            cardinalNumber = cardinalNumber,
            attendanceRate = stats.attendanceRate,
        )
    }

    fun toStats(attendances: List<Attendance>): ClubAttendanceStats =
        ClubAttendanceStats.fromCounts(
            attendances.count { it.status == AttendanceStatus.ATTEND },
            attendances.count { it.status == AttendanceStatus.ABSENT },
        )

    fun toResponse(attendance: Attendance): AttendanceResponse =
        AttendanceResponse(
            id = attendance.id,
            status = attendance.status,
            title = attendance.session.title,
            start = attendance.session.start,
            end = attendance.session.end,
            location = attendance.session.location,
        )

    fun toInfoResponse(attendance: Attendance): AttendanceInfoResponse =
        AttendanceInfoResponse(
            id = attendance.id,
            status = attendance.status,
            name = attendance.clubMember.user.name,
            department = attendance.clubMember.user.department,
            studentId = attendance.clubMember.user.studentId,
            memberStatus = attendance.clubMember.memberStatus,
        )

    fun toQrTokenResponse(
        session: Session,
        expiredAt: LocalDateTime,
    ): QrTokenResponse =
        QrTokenResponse(
            sessionId = session.id,
            code = session.code,
            expiredAt = expiredAt,
        )
}
